File: Engine/src/main/java/org/etieskrill/engine/graphics/OrthographicCamera.java
package org.etieskrill.engine.graphics;

import org.joml.Matrix4d;
import org.joml.Matrix4f;
import org.joml.Vector2f;
import org.joml.Vector3f;

public class OrthographicCamera extends Camera {
    
    private final Vector2f size;
    private float top, bottom, left, right;
    
    public OrthographicCamera(Vector2f size, Vector3f origin) {
        super();
        this.size = new Vector2f();
        setSize(size);
        setPosition(origin);
        setPerspective(new Matrix4f().ortho(left, right, bottom, top, near, far));

        if (autoUpdate) update();
    }

    public OrthographicCamera(Vector2f size) {
        this(size, new Vector3f(0f));
    }

    @Override
    protected void updatePerspective() {
        //TODO proper zoom
        float zoom = 1f / this.zoom;
        perspective.setOrtho(zoom * left, zoom * right, zoom * bottom, zoom * top, zoom * near, zoom * far);
    }

    @Override
    public Camera setPosition(Vector3f position) {
        this.position.set(position);
        updateDimensions();
        if (autoUpdate) update();
        return this;
    }

    public void setSize(Vector2f size) {
        this.size.set(size);
        updateDimensions();
        if (autoUpdate) update();
    }

    private void updateDimensions() {
        right = 0.5f * size.x();
        left = -right;
        bottom = 0.5f * size.y();
        top = -bottom;
    }
    
}

File: Engine/src/commonMain/kotlin/org/etieskrill/engine/graphics/camera/OrthographicCamera.kt
package org.etieskrill.engine.graphics.camera

import org.joml.Matrix4f
import org.joml.Vector2ic

class OrthographicCamera private constructor(
    viewportSize: Vector2ic,
    private var top: Float,
    private var bottom: Float,
    private var left: Float,
    private var right: Float,
    private val manualViewportSize: Boolean
) : Camera(viewportSize) {

    constructor(viewportSize: Vector2ic) : this(
        viewportSize,
        -.5f * viewportSize.y(), .5f * viewportSize.y(),
        -.5f * viewportSize.x(), .5f * viewportSize.x(),
        false
    )

    constructor(viewport: Vector2ic, top: Float, bottom: Float, left: Float, right: Float)
            : this(viewport, top, bottom, left, right, true)

    init {
        dirty = true
        update()
    }

    override fun updateViewportSize() {
        if (manualViewportSize) return

        top = -0.5f * viewportSize.y()
        bottom = 0.5f * viewportSize.y()
        left = -0.5f * viewportSize.x()
        right = 0.5f * viewportSize.x()
    }

    override fun updateProjection() {
        //TODO proper zoom
        val zoom = 1f / this.zoom
        (projection as Matrix4f).setOrtho(zoom * left, zoom * right, zoom * bottom, zoom * top, zoom * near, zoom * far)
    }

}
